### DISCUSSION TITLE
Adding dependecies to the peggy-loader

### DISCUSSION THREAD

**Discussion by s5b** [RocketChat/fuselage#1365] [Q&A]
I have a `peggy.js` grammar that has a dependency on [big.js](https://github.com/MikeMcl/big.js/) for managing numeric arithmetic and comparisons as part of the implementation. I'm currently passing the `Big` function through the runtime `options` parameter to the **_generated_** parser, but this is annoying and requires a lot of options setup code (noise) every time I want to use the parser in my production code.

I want to generate a self-contained parser that has the dependency built in.

I see that there is a `--dependency` CLI attribute to have the parser _require_ a specified dependency. And if I understand this correctly, this would generate a parser that is more _stand-alone_ than the _runtime options configuration_ that I'm now using.

My problem is that I don't know how to configure this dependency management as part of my `webpack` build. In the `webpack.config.js` file I have the following configuration:

```JavaScript
module.exports = {
  // ...
  module: {
    rules: [
    // ...
    {test: /\.peggy$/i, loader: 'peggy-loader'}
    // ...
```

I'm  hoping that I can add the _dependencies_ reference to the end of the `loader` definition. But I'm not sure of the syntax. I expect it to be something like: `{test: /\.peggy$/i, loader: 'peggy-loader?dependency=Big:big.js'}`

And then just reference the dependency in the grammar as, for example, `const bigTwentyThree = Big('23');`

However, my expectation attempt for the `webpack` configuration above doesn't work. 

**Can somebody provide me with some guidance on how to use the loader to include my dependency, please?**

Thanks for reading.

**Comment by ggazzo** [comment]
have you tested: 

``` 
{
test: /\.peggy$/i, use:{ loader: 'peggy-loader', options: { dependencies: {Big: 'big.js'} }
}
```
 to be very honest we never tested this options  settings, if you can, open a pull request creating a extra package reproducing your case, so we can work on it

**Comment by ggazzo** [comment]
@s5b can you check if this would solve your issue?

https://github.com/RocketChat/fuselage/pull/1368

**Comment by s5b** [comment]
@ggazzo Thank you for replying; very, very much appreciated.

I tried something similar to your example above, but without success. I tried the syntax: `peggy-loader?dependencies={Big:'big.js'}`

I'll have another go with the alternative that you're demonstrating above. 

And thanks for the reference to #1368, too. That will be helpful.

**Reply by ggazzo** [reply]
the pull request just accepts the key/value without `dependencies`.

also I tested this
> test: /\.peggy$/i, use:{ loader: 'peggy-loader', options: { dependencies: {Big: 'big.js'} }

and it it inserts a const Big = require('big.js') 

please share some log or gist if you can

**Comment by s5b** [comment]
**NOTE: I've solved this `webpack` configuration error - on my part. Will continue with the _dependencies_ investigation.**


@ggazzo I'm having trouble configuring `peggy-loader` with `webpack` (5.91.0). 

Rather than install `peggy` and `@rocket-chat/peggy-loader` as _production_ dependencies (as indicated in the project README), I have installed them using the following commands as `dev` dependencies:

```bash
npm install --save-dev peggy
npm install --save-dev @rocket.chat/peggy-loader
```

Following your recommendation, I changed my `webpack` configuration to be:

```
{test: /\.pegjs$/i, use:{ loader: 'peggy-loader', options: { dependencies: {Big: 'big.js'} }}}
```

This configuration gave me the error that `Can't resolve 'peggy-loader' ...`. So I changed the configuration to be:

```
{test: /\.pegjs$/i, use:{ loader: '@rocket-chat/peggy-loader', options: { dependencies: {Big: 'big.js'} }}}
```

And this gives me the error `Can't resolve '@rocket-chat/peggy-loader' …`. So, it's not processing my grammar at all.

There doesn't appear to be a top-level entry for `peggy-loader` in my `./node-modules/` directory.

Can you please tell me what I'm doing wrong here? (Edit: My bad!)

**Comment by s5b** [comment]
@ggazzo The problem above was created by my own typographical error. 

I used `@rocket-chat/peggy-loader` and it should be `@rocket.chat/peggy-loader`; _hyphen_ should be _dot_.

**Please ignore my most recent message.**
